sockeye/inference.py

```diff
@@ -511,35 +529,145 @@ def get_max_output_length(input_length: int):
+            yield TranslatorInput(sentence_id=self.sentence_id,
+                                  tokens=self.tokens[i:i + chunk_size],
+                                  factors=factors,
+                                  chunk_id=chunk_id)
+
+
+def make_input_from_plain_string(sentence_id: int, string: str) -> TranslatorInput:
+    """
+    Returns a TranslatorInput object from a plain string.
+
+    :param sentence_id: An integer id.
+    :param string: An input string.
+    :return: A TranslatorInput.
+    """
+    return TranslatorInput(sentence_id, tokens=list(data_io.get_tokens(string)), factors=None)
+
+
+def make_input_from_json_string(sentence_id: int, json_string: str) -> TranslatorInput:
+    """
+    Returns a TranslatorInput object from a JSON object, serialized as a string.
+
+    :param sentence_id: An integer id.
+    :param json_string: A JSON object serialized as a string.
+    :return: A TranslatorInput.
+    """
+    jobj = json.loads(json_string, encoding=C.JSON_ENCODING)
+    tokens = jobj.get(C.JSON_TEXT_KEY)
+    if tokens is None:
+        raise ValueError("Missing 'text' field in JSON string: %s", json_string)
+
+    tokens = list(data_io.get_tokens(tokens))
+    factors = jobj.get(C.JSON_FACTORS_KEY)
+    if isinstance(factors, list):
+        factors = [list(data_io.get_tokens(factor)) for factor in factors]
+    else:
+        factors = None
+
+    return TranslatorInput(sentence_id=sentence_id, tokens=tokens, factors=factors)
+
+
+def make_input_from_factored_string(sentence_id: int,
+                                    factored_string: str,
+                                    translator: 'Translator',
+                                    delimiter: str = C.DEFAULT_FACTOR_DELIMITER) -> TranslatorInput:
+    """
+    Returns a TranslatorInput object from a string with factor annotations on a token level, separated by delimiter.
+    If translator does not require any source factors, the string is parsed as a plain token string.
+
+    :param sentence_id: An integer id.
+    :param factored_string: An input string with additional factors per token, separated by delimiter.
+    :param translator: A translator object.
+    :param delimiter: A factor delimiter. Default: '|'.
+    :return: A TranslatorInput.
+    """
+    utils.check_condition(bool(delimiter) and not delimiter.isspace(),
+                          "Factor delimiter can not be whitespace or empty.")
+
+    model_num_source_factors = translator.num_source_factors
+
+    if model_num_source_factors == 1:
+        return make_input_from_plain_string(sentence_id=sentence_id, string=factored_string)
+
+    tokens = []  # type: Tokens
+    factors = [[] for _ in range(model_num_source_factors - 1)]  # type: List[Tokens]
+    for token_id, token in enumerate(data_io.get_tokens(factored_string)):
+        pieces = token.split(delimiter)
+        utils.check_condition(all(pieces) and len(pieces) == model_num_source_factors,
```

Comment:
So far no input to inference could cause the sockeye process to crash, e.g. empty inputs were just returned. All words are mapped to some word id (<unk> if nothing matches). I really think that we should continue to do so. Specifically for the factors I would very much prefer to fill factors with <unk> rather than raising an exception here. In order to make the user aware we should log to stderr that we failed parsing the input. This way it is not possible for a single wrongly formatted input to crash the process. For training this is a different story.